rfcs/text/0011_global_search.md

```diff
@@ -0,0 +1,693 @@
+- Start Date: 2020-04-19
+- RFC PR: [#64284](https://github.com/elastic/kibana/pull/64284)
+- Kibana Issue: [#61657](https://github.com/elastic/kibana/issues/61657)
+
+# Summary
+
+A core API exposed on both public and server sides, allowing to search for various objects and
+register result providers.
+
+# Basic example
+
+- registering a result provider:
+
+```ts
+coreSetup.globalSearch.registerResultProvider({
+  id: 'my_provider',
+  find: (term, options, context) => {
+    const resultPromise = myService.search(term, context.core.savedObjects.client);
+    return from(resultPromise);
+  },
+});
+```
+
+- using the `find` API from the client-side:
+
+```ts
+coreStart.globalSearch.find('some term').subscribe(({ results, complete }) => {
+    showAsyncSearchIndicator(!complete);
+    updateResults(results);
+});
+```
+
+# Motivation
+
+Kibana should do it's best to assist users searching for and navigating to the various objects present on the Kibana platform.
+
+We should expose an API to make it possible for plugins to search for the various objects present on a Kibana instance.
+
+The first consumer of this API will be the global search bar [#57576](https://github.com/elastic/kibana/issues/57576). This API
+should still be generic to answer similar needs from any other consumer, either client or server side.
+
+# Detailed design
+
+## API Design
+
+### Result provider API
+
+#### common types
+
+```ts
+/**
+ * Static list of all the possible result types.
+ * Ordinal value matters here, as it will be used to sort results of different types.
+ */
+enum SEARCH_TYPE {
+  // non-exhaustive
+  applications = 10,
+  dashboard = 20,
+  visualization = 30,
+  search = 40,
+}
+
+/** @public */
+type GlobalSearchResultType = keyof typeof SEARCH_TYPE;
+
+/**
+ * Options provided to {@link GlobalSearchResultProvider | result providers} `find` method.
+ * Currently empty and only present to keep the API future-proof.
+ */
+interface GlobalSearchOptions {}
+
+/**
+ * Representation of a result returned by a {@linlkGlobalSearchResultProvider | result provider}
+ */
+interface GlobalSearchResult {
+  /** an id that should be unique for an individual provider's results */
+  id: string;
+  /** the title/label of the result */
+  title: string;
+  /** the type of result */
+  type: GlobalSearchResultType;
+  /** an optional EUI icon name to associate with the search result. If not specified, will use a default icon */
+  icon?: string;
+  /** The url to navigate to this result. This can be either an absolute url, or a path relative to the performing request's basePath */
+  url: string;
+  /** the score of the result, from 1 (lowest) to 100 (highest) */
+  score: number;
+  /** an optional record of metadata for this result */
+  meta?: Record<string, any>;
```

Comment:
Do we have any known usages for this?

Reply:
++ I had the same question. Was also wondering if the meta could really be `Record<string, any>` or if we would expect it to be serializable.

Reply:
> Do we have any known usages for this?

We don't. TBH I'm just futur proofing the API. Potential use would be to add arbitrary data that would be usable for the consumer for UI representation.

If you think this is unnecessary I can remove it.

> or if we would expect it to be serializable

We do. I will change `any` with something more restrictive. Do we already have an type representing `Serializable` data?

Reply:
> We don't. TBH I'm just futur proofing the API. Potential use would be to add arbitrary data that would be usable for the consumer for UI representation.
> 
> If you think this is unnecessary I can remove it.

Less code usually means less bugs, but since we're just passing this along, I don't have any problems with keeping it, especially if we can imagine some likely use cases.

Reply:
> Do we already have an type representing Serializable data?

@pgayvallet  The closest thing I can think of is `SavedObjectAttributes`, however that interface has its own issues -- there is a thread on the topic with some proposed alternatives which I will DM you.

Reply:
Is 

```ts
type Serializable = string | number | boolean | PrimitiveArray | PrimitiveRecord;
// interfaces trick for self reference
interface PrimitiveArray extends Array<Serializable> {}
interface PrimitiveRecord extends Record<string, Serializable> {}
```

too naive of an implementation?

Reply:
Added in https://github.com/elastic/kibana/pull/64284/commits/5d6e4cbbb48a8ec2155fcd51f08b8198a3aee9bc

Reply:
Might need `null` or `undefined` in there too? Otherwise that makes sense to me.